experimental/dds/tree2/src/simple-tree/proxies.ts

```diff
@@ -1011,6 +1067,46 @@ function extractContentObject<T extends object>(input: T): ExtractedFactoryConte
 	};
 }
 
+/**
+ * If `input` is a factory-created proxy, this will return the content stored in that proxy and forward the `typeNameSymbol` on the proxy to `object`.
+ * Otherwise, it will just return `input`.
+ */
+function extractContentIfProxy<T>(input: T): {
+	content: T;
+	type: "object" | "list" | "map" | "leaf";
+	fromFactory: boolean;
+} {
+	let content = input;
+	let fromFactory = false;
+	const rawEditNode = tryGetEditNode(input);
+	if (rawEditNode !== undefined) {
+		const factoryContent = extractRawNodeContent(rawEditNode);
+		if (factoryContent === undefined) {
+			// We were passed a proxy, but that proxy doesn't have any raw content.
+			throw new Error("Cannot insert a node that is already in the tree");
+		}
+		content = factoryContent as T;
+		fromFactory = true;
+	}
+
+	// TODO: This checking could be made stronger by using `class-tree/Tree.schema(content).kind`
+	// to get the type when `content` is a proxy. At the moment, importing `Tree.schema` causes a cycle.
+	let type: "object" | "list" | "map" | "leaf";
+	if (isFluidHandle(content)) {
+		type = "leaf";
+	} else if (isReadonlyArray(content)) {
+		type = "list";
+	} else if (content instanceof Map) {
+		type = "map";
+	} else if (content !== null && typeof content === "object") {
+		type = "object";
+	} else {
+		type = "leaf";
```

Comment:
Nit: should we assert anything here?

Reply:
What do you have in mind?

Reply:
Sorry, I was originally thinking there might be value in asserting the content was of the shape we expect, since this is the catch-all else case. But after reading the code again, this _has_ to be a primitive or undefined. So disregard 😋